# Generated by the protocol buffer compiler.  DO NOT EDIT!
# sources: error.proto
# plugin: python-betterproto
from dataclasses import dataclass

import betterproto


class WalletNotFound(betterproto.Enum):
    """When requesting for a specific wallet and it's not found"""

    WALLET_NOT_FOUND_UNKNOWN = 0
    WALLET_NOT_FOUND_ON_DEVICE = 1
    WALLET_NOT_FOUND_ON_CARD = 2


class WalletPartialState(betterproto.Enum):
    """When requesting for a specific wallet and it's in partial state"""

    WALLET_PARTIAL_STATE_UNKNOWN = 0
    WALLET_PARTIAL_STATE_LOCKED = 1
    WALLET_PARTIAL_STATE_DELETE = 2
    WALLET_PARTIAL_STATE_UNVERIFIED = 3
    WALLET_PARTIAL_STATE_OUT_OF_SYNC = 4


class CardError(betterproto.Enum):
    """
    When the card operation fails, one of these errors is communicated by
    device. For card specific errors, a card error Status word(SW) is returned,
    else `CARD_ERROR_UNKNOWN`.
    """

    CARD_ERROR_UNKNOWN = 0
    CARD_ERROR_NOT_PAIRED = 1
    CARD_ERROR_SW_INCOMPATIBLE_APPLET = 3
    CARD_ERROR_SW_NULL_POINTER_EXCEPTION = 4
    CARD_ERROR_SW_TRANSACTION_EXCEPTION = 5
    CARD_ERROR_SW_FILE_INVALID = 6
    CARD_ERROR_SW_SECURITY_CONDITIONS_NOT_SATISFIED = 7
    CARD_ERROR_SW_CONDITIONS_NOT_SATISFIED = 8
    CARD_ERROR_SW_WRONG_DATA = 9
    CARD_ERROR_SW_FILE_NOT_FOUND = 10
    CARD_ERROR_SW_RECORD_NOT_FOUND = 11
    CARD_ERROR_SW_FILE_FULL = 12
    CARD_ERROR_SW_CORRECT_LENGTH_00 = 13
    CARD_ERROR_SW_INVALID_INS = 14
    CARD_ERROR_SW_NOT_PAIRED = 15
    CARD_ERROR_SW_CRYPTO_EXCEPTION = 16
    CARD_ERROR_POW_SW_WALLET_LOCKED = 17
    CARD_ERROR_SW_INS_BLOCKED = 18
    CARD_ERROR_SW_OUT_OF_BOUNDARY = 19


class UserRejection(betterproto.Enum):
    """When user rejects the operation"""

    USER_REJECTION_UNKNOWN = 0
    USER_REJECTION_CONFIRMATION = 1


class DataFlow(betterproto.Enum):
    # protobuf decoding by nanopb api failed
    DATA_FLOW_DECODING_FAILED = 0
    # query received is: invalid or is of different app_function (eg. device auth
    # in between get logs)
    DATA_FLOW_INVALID_QUERY = 1
    # protobuf `optional` fields when checked on receiving end was absent.
    DATA_FLOW_FIELD_MISSING = 2
    # Wrong request is received in app_function. could be at start of flow or in-
    # between the flow
    DATA_FLOW_INVALID_REQUEST = 3
    # inactivity during wait-for-data from host. In-between a flow
    DATA_FLOW_INACTIVITY_TIMEOUT = 4
    # when the flow specific data validity fails. could be at start of flow or
    # in-between the flow
    DATA_FLOW_INVALID_DATA = 5
    # query received is: when app in unserviceable state (eg. export wallet
    # request before on-boarding, export wallet request on unauthenticated device
    # on main
    DATA_FLOW_QUERY_NOT_ALLOWED = 6


@dataclass
class CommonError(betterproto.Message):
    # ****** Protocol specific errors ******
    unknown_error: int = betterproto.uint32_field(1, group="error")
    corrupt_data: "DataFlow" = betterproto.enum_field(2, group="error")
    # The user needs to complete device setup
    device_setup_required: int = betterproto.uint32_field(4, group="error")
    # The specified wallet does not exist
    wallet_not_found: "WalletNotFound" = betterproto.enum_field(11, group="error")
    # The specified wallet is in partial state
    wallet_partial_state: "WalletPartialState" = betterproto.enum_field(
        12, group="error"
    )
    # ****** User action errors ****** Card Error
    card_error: "CardError" = betterproto.enum_field(21, group="error")
    # User rejection error
    user_rejection: "UserRejection" = betterproto.enum_field(22, group="error")
